compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp

```diff
@@ -727,134 +662,201 @@ checkDispatchForVectorDistribution(Operation *parentOp) {
     return failure();
   }
+  const SetVector<linalg::LinalgOp> &computeOps = maybeComputeOps.value();
+
+  const IREE::GPU::TargetWgpAttr wgp = target.getWgp();
+  const SmallVector<int64_t> bounds = root.getStaticLoopRanges();
 
   SmallVector<unsigned> parallelDims;
+  root.getParallelDims(parallelDims);
+
   SmallVector<unsigned> reductionDims;
-  op.getParallelDims(parallelDims);
-  op.getReductionDims(reductionDims);
+  root.getReductionDims(reductionDims);
+
+  const uint64_t lastReductionDim = reductionDims.back();
+  const bool lastReductionDimDynamic =
+      ShapedType::isDynamic(bounds[lastReductionDim]);
+  const int64_t lastReductionSize =
+      lastReductionDimDynamic ? kVectorDistributeReductionSizeToTargetIfDynamic
+                              : bounds[lastReductionDim];
+
+  const int64_t subgroupSize = target.getPreferredSubgroupSize();
+
+  // This logic to bail when the static reduction size is less than the
+  // subgroup size must be revisited. It is currently required to unblock
+  // a failing test. Moreover, if better performance is indeed obtained with
+  // another strategy, then that logic should be separated out, i.e. we must
+  // decouple the decision of if a strategy is valid from how performant it is.
+  if (!lastReductionDimDynamic && subgroupSize > lastReductionSize) {
+    return failure();
+  }
 
-  SmallVector<int64_t> bounds = op.getStaticLoopRanges();
-  IREE::GPU::TargetWgpAttr wgp = target.getWgp();
-  int64_t reductionSize = bounds[reductionDims.back()];
+  FailureOr<int64_t> maybeBitWidth = getBitWidth(root);
+  if (failed(maybeBitWidth)) {
+    return failure();
+  }
+  const int64_t bitWidth = maybeBitWidth.value();
 
-  if (ShapedType::isDynamic(reductionSize)) {
-    reductionSize = kVectorDistributeReductionSizeToTargetIfDynamic;
+  // Reduction distribution only supports 4/8/16/32 bit types now.
+  if (!llvm::is_contained({4, 8, 16, 32}, bitWidth)) {
+    return failure();
   }
 
-  bool hasDynamicReductionDim = false;
-  for (unsigned dim : reductionDims) {
-    if (ShapedType::isDynamic(bounds[dim])) {
-      hasDynamicReductionDim = true;
+  const unsigned largestLoadSizeInBits =
+      wgp.getMaxLoadInstructionBits().value_or(128);
+  const unsigned largestLoadSizeInElements = largestLoadSizeInBits / bitWidth;
+
+  // The current logic assumes that the reduction dimension is contiguous. In
+  // this case, we'd like threads to load a large number of elements. We
+  // try however to ensure threads either load a full vector, or nothing, to
+  // avoid masking which would cause serialized loads.
+  const unsigned threadLoads = [&]() {
+    if (lastReductionDimDynamic) {
+      return largestLoadSizeInElements;
     }
-  }
+    unsigned loads = largestLoadSizeInElements;
+    while (lastReductionSize % loads != 0 && loads > 1) {
+      loads /= 2;
+    }
+    return loads;
+  }();
+  assert(lastReductionDimDynamic || (lastReductionSize % threadLoads == 0));
 
-  int64_t subgroupSize = 0;
```

Comment:
This constraint that reductionSize % subgroupSize == 0 is removed in this PR.  TODO(newling) add tests (numerical and lit).